Add GIS information to job offers.

fullonic opened the pull request:
Related with: #196

Make possible generate coordinates from job offer address and display it on map.

To accomplish this functionality the following changes were made/introduced:
- Add a new app, `geolocation`, for all related geographic data.
- A new `Address` model is introduced to deal with job address information.
- The old `location` field from `Job` model is replaced by a new one, which makes a relation to the Address model. [1]
- Replace location field of **new job creation form page** by a more detailed address fields.
- Add a map to display all jobs offers. [2]
- ~~Update POST `api/jobs` to be able to send the new address information.~~
- Add a background task (with celery) to call the geocoder service in order to get the job address coordinates.
- As database image, we are now using [kartoza/postgis:13.0](https://hub.docker.com/r/kartoza/postgis/).

There are still things that can be improved or changed! Ideas and suggestions are welcome.
Right now, for every job post, we ask to add the job address. However it could be done as well when the employer signup and save it as default address. This would allows us to "suggest" an address when a employer add new job by automatically populate the address fields.

Also, maybe some things can be better named. For instance, the Address models fields name are named in the same way as are on [OSM Nominatim](https://nominatim.org/release-docs/develop/api/Search/#parameters), but can be change.

What information should we include about the job on the map? Currently shows the "company name" and the "total of opening positions".

~~Currently there is not support to edit address on the edit job offer.~~

Doubts:
- Is the current way of deal with nested models okay? Any other recommendation? I couldn't find one from django documentation.

Things that are still missing:
- A helper function to automatically generate an address entry to the current jobs in our database.
- A test fixture to show data on map.
- Run `makemessages`.
- Add `Address` model to admin panel.

[1] Currently named as “address”, which could be replaced later by location.
The location field is still present on the Job model but should be deleted once we add the address information to the current offers that exist in our database.
[2] There is a temporary hyperlink to the map page on the nav bar. Ideas about how and where display the map url and style it are needed.

fullonic committed:
Refractor JobsViewList create method.

Instead of using JobsViewList.create() method, now we use
JobsViewList.perform_create which allows
us to pass a user propriety as readonly to
the job serializer.

fullonic committed:
Refactor jobs tests

fullonic committed:
Use gettext instead of the deprecated ugettext

fullonic committed:
Add geolocation app

fullonic committed:
Add GeoCoder

Convert address to coordinates using geopy and nominatim

fullonic committed:
Add model Address to save geolocation information.

fullonic committed:
Change docker postgres image.

We move from official postgres image to kartoza/postgis to
get  PostGIS support.

fullonic committed:
Add PostGIS settings.

Add gdal library to docker image

fullonic committed:
Add tests to Address model

fullonic committed:
Allows pass Address.full_address to geocoder

Changes GeoCoder.get_coordinates() query params.
Update tests.

fullonic committed:
Fix missing set geo_point when adding coordinates to address.

fullonic committed:
Add custom query manager.

Table records can be converted to geojson using .geojson() on querysets.

fullonic committed:
Run geocoder as celery task.

In order to get the coordinates for a new
address, we run the operations as a background task.

fullonic committed:
Add coordinates information on new entry creation.

fullonic committed:
Add relation Job model with  Address model.

A temporary field "geo_location" was added to Job Model.
It may later replace the current location field.

fullonic committed:
Add custom GeoJSON serializer.

Now is possible add job and company
information to 'feature' 'proprieties' field
inside the generated geojson.

fullonic committed:
Set queue to celery task

fullonic committed:
Add Address Serializer

fullonic committed:
Get geo coordinates when calling Address.save()

fullonic committed:
Add temporary classmethod save_job_with_address().

Allows savea address when adding a new job entry.

fullonic committed:
Add relation of Address and Job serializer.

Add also a create() to JobSerializer to allow writable nested serializer

fullonic committed:
Create map page.

- Add a rest endpoint to get all jobs in geojson format
- Add an url to show the map on browser.
- Create map template, js and css
- Add basic test

- Update jobsapp tests

fullonic committed:
Build geojson without serializer.

Fix issues with coordinates order and srid bad conversion.

fullonic committed:
Handle better when coordinates are not found.

fullonic committed:
Add address form job form view.

Now is possible add a more detailed information about job location.

fullonic committed:
Style address form fields.

Also removes the old location field.

fullonic committed:
Improves readability and semantics.

Add Map class as proxy to Address model.
Now instead of Address.objects.geojson() we do Map.objects.geojson().

fullonic committed:
Add default value to "old" location field.

fullonic committed:
Clean code.

Remove unused code.
Remove some comments.

fullonic committed:
Remove location field from jobsapp tests.

fullonic committed:
Update docker test images.

fullonic committed:
Update location information on job detail page.

fullonic committed:
undo gettext changes

fullonic committed:
Revert "Use gettext instead of the deprecated ugettext"

This reverts commit b7d376f55dc5e5a5caf9f18c6267f1f20dfc6809.

fullonic committed:
Fix  geojson serializer coordinates convertion issue.

fullonic committed:
Fix SRID typo

fullonic committed:
Remove number field from Address model.

fullonic committed:
Improve code.

Simplify job form_valid().
Test against not found address.
Remove duplicate tests.

fullonic committed:
Allow edit address inside update job page.

fullonic committed:
Format code and pass all tests..

fullonic committed:
Allows user get their current location on map.

Add locate-control plugin to map.

fullonic committed:
Ensures that there is only an address with the same coordinates.

fullonic committed:
Remove filled positions from geojson response.

Now only the unfilled are displayed on map.

fullonic committed:
Format code.

fullonic committed:
Add relation beetween user and addres models

fullonic committed:
Only one company is displayed on map per coordinate

Due the way that geojson data is displayed on map, if there is more than
one geometry point with the same coordinates, only one will be show.

To avoid this problem, we must ensure that:
1 - Every job offer from the same company with the same location receive
the same coordinates.
2 - We offset duplicated coordinates from different companies.

1. We add a database level constrain to ensure that one company have
only one address with the same coordinates.
Raising an 'IntegrityError' (due te above constrain), we know that the
current job offer company already have that address. Therefore we just
need to add the new job to the database address.
If the address doesn't exist, we create a new one and set it to the new
job offer.

At the end of the process, we are sure that there is only one address
per company-coordinates combination. So, every new offer that match this
combination will be add to the current db address. When a company have
more than one job offers 'unfilled', that information will be present on
company map popup as "open positions".
The described logic is made at 'geolocation.geo_utils.add_address_to_job'

2. To be able to display companies with the same coordinates, we
generate a small offset to the coordinates while generating the geojson
data. This create a small performance penalty as we need to check all
coordinates but could be solved by caching the generated geojson. [TODO]
So when we run Map.objects.geojson(), we filter on fly the unfilled
positions, the companies without jobs, and offset duplicated coordinates

fullonic committed:
Remove unused code

fullonic committed:
Refactor code

- Pass all tests
- Minor improves on tests
- Improve logic to deal with jobs submissions
- Change map icon

fullonic committed:
Fix broken test

fullonic committed:
Improve manager geojson query

Rename as well check duplicated coordinates.

fullonic committed:
Fix some solarlint warnings

fullonic committed:
Code refactoring

Split logic of add_address_to_job into smaller functions and
model methods.
Move view query to get_queryset method. It avoids raise a error if there is
any data on the database.

fullonic committed:
Simplify logic to create an address for a job

XaviTorello commented on geolocation/migrations/0002_auto_20210125_1455.py:
```diff
@@ -0,0 +1,28 @@
+            },
+            bases=('geolocation.address',),
+        ),
+        migrations.RemoveField(
```
Maybe it's better to keep (when possible) just one unique migration per PR

XaviTorello commented on geolocation/models.py:
```diff
@@ -0,0 +1,106 @@
+        return ", ".join(
+            [
+                field
+                for field in (self.street, self.city, self.county, self.country)
```
maybe it's easier to identify the `state` instead of the `county`, but as you will :)

XaviTorello reviewed (changes requested):
Wow! Amazing job, many thanks @fullonic 

In general looks good to me, but there are some core decisions to be confirmed with the development team. I.e, which is more important, the address fields or the geolocated position? we want to display a map in the job offer view? addresses instances should be reused by `user` or not? ...

I propose you to review it together, and clarify the core definitions to go further with this PR

What do you think?

fullonic committed:
Improve tests

Add mock to some geocode calls
Fix filter issue because was missing user kwarg

fullonic committed:
Save address at the same time that job offer

Instead of saving the address information
on a background task, we now save it at the
same time than we save the job offer.
Also:
    - Add tests for HTTP job form create and update offer
    - Clean up some tests

fullonic committed:
Remove unsued code from Address model

Add state to Address.__str__

fullonic committed:
Merge migrations

fullonic committed:
Clean code

Remove unused imports
Remove unused code
Mock some tests to have a faster test suite

Extras:
Add pytest-mock to requirements-test.txt
Add rate limiter to the  geocode service

fullonic committed:
Merge branch 'master' into 196_activate_gis_information

fullonic committed:
Remove duplicated code on form views

Add a new functiont o avoide duplicate the same logic in two form views.

Add tests against invalid form submition

fullonic committed:
Remove lat and lon from databse model
